test/metabase/test/data.clj

```diff
@@ -212,3 +213,48 @@
   `(-with-temp-db ~database-definition
      (fn [~db-binding]
        ~@body)))
+
+
+(defmacro expect-with-test-db
+  "Identical to `expect` but wraps the entire evaluation in `with-temp-db` so that both expected and actual statements
+   have access to the test data required for the test.
+
+     (expect-with-test-db [_ test-data]
+       (id :venues :id)
+       (sel :one :field [ForeignKey :destination_id] :origin_id (id :checkins :venue_id)))"
+  [test-db-binding expected actual]
+  (let [fn-name (gensym)]
+    `(def ~(vary-meta fn-name assoc :expectation true)
+       (fn []
+         (with-temp-db ~test-db-binding
+           (expectations/doexpect ~expected ~actual))))))
+
+
+;;; ## ---------------------------------------- Test-Data Interface ----------------------------------------
+
+
+(defn with-test-data [i-test-data f]
+  (try
+    (binding [*sel-disable-logging* true]
+      (when (fn? (:before i-test-data))
+        ((:before i-test-data)))
+      (binding [*sel-disable-logging* false]
+        (f)))
+    (finally
+      (binding [*sel-disable-logging* true]
+        (when (fn? (:after i-test-data))
+          ((:after i-test-data)))))))
```

Comment:
I would write this as 
```clojure
(defn with-test-data [{:keys [before after]} i-test-data f]
  (try
    (when (fn? before)
      (binding [*sel-disable-logging* true]
        (before)))
    (f)
    (finally
      (when (fn? after)
        (binding [*sel-disable-logging* true]
          (after))))))
```

*  Setting up bindings are a bit expensive what with all the thread-safety and try-finallys going on. It looks like in the majority of cases these bindings aren't going to be used at all :disappointed: . Seeing how this code could easily go either way, I would vote on nesting them inside the `whens`. This way they're only being established in cases where they're being used
*  If you think about it, it doesn't make sense to establish a binding around a pair of forms, and then establish a binding in the second to reverse the original. :unamused: 
*  I think this code is upwards of 50% more readable if you destructure `i-test-data` so you can just access `before` and `after` directly :flushed: